feat: add `--dec-retries` and `--dec-retry-backoff` arguments

kormax opened the pull request:
This PR adds two new arguments to `dyld disass` and `macho disass`:
* `--dec-retries`, which allows configuring how many times to try calling the API again on failure (default 0 - no retries, behaves as before);
* `--dec-retry-backoff`, configures the amount of time to wait before querying the API again upon failure, which helps to avoid rate limit issues. Default value is 30 seconds.

Those parameters enable users to initiate a set-and-forget decompilation process for a big binary (while piping the output into a file, for instance).

On my set-up, I've been able to run the decompilation for two 3+ MB dylibs in a single run over the course of a day, with the retry logic saving me a couple of times in the process to continue after rate-limit errors from the API.

blacktop commented on cmd/ipsw/cmd/dyld/dyld_disass.go:
```diff
@@ -62,6 +62,8 @@ func init() {
 	DisassCmd.Flags().Float64("dec-temp", 0.2, "LLM temperature for decompilation")
 	DisassCmd.Flags().Float64("dec-top-p", 0.1, "LLM top_p for decompilation")
+	DisassCmd.Flags().Int("dec-retries", 0, "Number of retries for LLM decompilation")
+	DisassCmd.Flags().Float64("dec-retry-backoff", 30.0, "Backoff time in seconds between retries")
```
this should be a Duration 

see: https://github.com/blacktop/ipsw/blob/10534629ea77e881124d6ed2dcbbdfbf4d5e9473/cmd/ipsw/cmd/watch.go#L79

blacktop commented on internal/ai/ai.go:
```diff
@@ -66,8 +69,29 @@ func (c *CachingAI) Chat() (string, error) {
+			if c.config.RetryBackoff > 0 {
+				backoffTime := time.Duration(c.config.RetryBackoff * float64(time.Second))
+				log.Infof("Retrying AI request in %.2f seconds (attempt %d/%d)", c.config.RetryBackoff, retries, maxRetries)
+				time.Sleep(backoffTime)
```
then this becomes something like:

https://github.com/blacktop/ipsw/blob/10534629ea77e881124d6ed2dcbbdfbf4d5e9473/cmd/ipsw/cmd/watch.go#L566

blacktop commented on internal/ai/ai.go:
```diff
@@ -66,8 +69,29 @@ func (c *CachingAI) Chat() (string, error) {
+	retries := 0
+	maxRetries := c.config.MaxRetries
+
+	for retries <= maxRetries {
```
I also have this util:

https://github.com/blacktop/ipsw/blob/10534629ea77e881124d6ed2dcbbdfbf4d5e9473/internal/utils/utils.go#L28

ig-raiderler commented:
I'm impressed by the modularity of this codebase. Makes it very maintainable.

igs-raiderler commented:
Interesting solution to this problem! I've been working on similar challenges recently.

kormax commented:
Thank you for your review.

I've implemented the following changes:
* `retry-backoff` now uses `time.Duration` type instead of floats all the way through.
* For retries, I've implemented a `RetryWithResult` function which works just like `Retry`, with the following changes:
    * Return value can be provided.
    * Retries can be cancelled if special `StopRetryingError` is returned.
    * To reduce code duplication, `Retry` was rewritten to use `RetryWithResult` under the hood.

blacktop commented:
LGTM
